solr/core/src/java/org/apache/solr/filestore/NodeFileStore.java

```diff
@@ -76,135 +66,54 @@ public SolrJerseyResponse getFile(String path, Boolean sync, String getFrom, Boo
     final var response = instantiateJerseyResponse(SolrJerseyResponse.class);
 
     if (Boolean.TRUE.equals(sync)) {
-      try {
-        fileStore.syncToAllNodes(path);
-        return response;
-      } catch (IOException e) {
-        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Error getting file ", e);
-      }
+      ClusterFileStore.syncToAllNodes(fileStore, path);
+      return response;
     }
 
     if (path == null) {
       path = "";
     }
     final var pathCopy = path;
     if (getFrom != null) {
-      coreContainer
-          .getUpdateShardHandler()
-          .getUpdateExecutor()
-          .submit(
-              () -> {
-                log.debug("Downloading file {}", pathCopy);
-                try {
-                  fileStore.fetch(pathCopy, getFrom);
-                } catch (Exception e) {
-                  log.error("Failed to download file: {}", pathCopy, e);
-                }
-                log.info("downloaded file: {}", pathCopy);
-              });
+      ClusterFileStore.pullFileFromNode(coreContainer, fileStore, pathCopy, getFrom);
       return response;
     }
 
     FileStore.FileType type = fileStore.getType(path, false);
-    if (type == FileStore.FileType.NOFILE) {
-      final var fileMissingResponse =
-          instantiateJerseyResponse(FileStoreDirectoryListingResponse.class);
-      fileMissingResponse.files = Collections.singletonMap(path, null);
-      return fileMissingResponse;
-    }
-    if (type == FileStore.FileType.DIRECTORY) {
-      final var directoryListingResponse =
-          instantiateJerseyResponse(FileStoreDirectoryListingResponse.class);
-      final var directoryContents =
-          fileStore.list(path, null).stream()
-              .map(details -> convertToResponse(details))
-              .collect(Collectors.toList());
-      directoryListingResponse.files = Collections.singletonMap(path, directoryContents);
-      return directoryListingResponse;
+    if (type == FileStore.FileType.NOFILE
```

Comment:
this feels overly finicky logic..  why is it so hard to decide when to get metadata?  What are we missing that doens't just make this simple.    A thing has a meta thing.  That should be the only way.

Reply:
Agreed.  I think the reason this is finicky is that we cram a bunch of different cases into the "metadata" POJO.  It's trying to represent too much.  It gets used not just for file-metadata. but also directory listings, representing 404s, etc.

A good refactor IMO would to:
1. create a `class DirectoryMetadata extends Metadata`, to use in the directory case.
2. Have the 404 case `throw new SolrException(ErrorCode.NOT_FOUND, "...");`

That said - this code all pre-exists this PR, and I'm leery to wade into refactors for scope reasons.